File: cpp/examples/linux/humidity/DHT11/complete/main.cpp
#include <cstdio>
#include <unistd.h>
#include "DHTxxTransportLinux.h"
#include "DHT11.h"

int main() {
    const char* chip_path = getenv("GPIO_CHIP") ? getenv("GPIO_CHIP") : "/dev/gpiochip0";
    int line_num = getenv("GPIO_LINE") ? atoi(getenv("GPIO_LINE")) : 4;
    DHTxxTransportLinux transport(chip_path, (unsigned)line_num);

    DHT11Full dht(transport, 3);                                           // Create DHT11 driver, (transport, max_retries=3)

    float t = dht.read_temperature();                                      // Read temperature, () → float °C
    float h = dht.read_humidity();                                         // Read humidity, () → float %RH
    printf("t=%.1f h=%.1f\n", t, h);
    float t2, h2;
    bool ok = dht.read_retry(5, t2, h2);                                   // Read with retries, (max_retries 1..255, t out, h out) → bool ok
    uint8_t raw[5];
    dht.read_raw_with_retry(raw);                                          // Read raw frame, (out[5]) → bool ok
    printf("retry_ok=%d raw[0]=0x%02X\n", ok, raw[0]);
    return 0;
}

File: rust/examples/linux/humidity/dht11/complete/src/main.rs
use gpio_cdev::{Chip, LineRequestFlags};
use linux_embedded_hal::CdevPin;
use periph::chips::humidity::Dht11Full;
use periph::transport::dhtxx::DHTxxTransportLinux;

fn main() {
    let chip_path   = std::env::var("GPIO_CHIP").unwrap_or_else(|_| "/dev/gpiochip0".into());
    let line_offset: u32 = std::env::var("DHT11_LINE").ok().and_then(|v| v.parse().ok()).unwrap_or(4);

    let mut chip = Chip::new(&chip_path).expect("open gpio chip");
    let handle = chip.get_line(line_offset).expect("get data line")
        .request(LineRequestFlags::INPUT, 0, "dht11_complete").expect("request line");
    let pin = CdevPin::new(handle).expect("dht11 pin");

    let transport = DHTxxTransportLinux::new(pin);
    let mut dht = Dht11Full::new(transport, 3);  // Create DHT11 driver, (transport, max_retries=3)

    let t = dht.read_temperature().unwrap_or(0.0);  // Read temperature, () → f32 °C
                                                    // returns a fresh conversion each call
    let h = dht.read_humidity().unwrap_or(0.0);     // Read humidity, () → f32 %RH
                                                    // returns a fresh conversion each call
    let r = dht.read_retry(5);                      // Read with retries, (max_retries=5) → (f32 °C, f32 %RH)
                                                    // retries up to 5 times on checksum error
    let raw = dht.read_raw().unwrap_or([0; 5]);     // Read raw frame, () → [u8; 5]
                                                    // returns the validated 5-byte frame
    println!("t={} h={} retry_t={} raw[0]=0x{:02X}", t, h, r.map(|(t, _)| t).unwrap_or(0.0), raw[0]);
    println!("===DONE: 0 passed, 0 failed===");
}
